{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Unit testy v Pythonu\n",
    "## Proč psát testy\n",
    "Představme si, že jsme napsali nějaký kód, dejme tomu funkci provádějící určitý výpočet. Chceme-li vědět, zda funguje, obvykle ho ručně otestujeme. Tj. danou funkci provoláme s určitými parametry a poté zkontrolujeme, zda výsledek odpovídá našim očekáváním. Pakliže ano, tak jsme šťastní, že nemusíme nic řešit. Pokud ale na výstupu funkce obdržíme nesmysl, musíme provést opravu.  \n",
    "Představme si nyní, že bychom naši funkci otestovali v rámci několika scénářů. Například bychom implementovali dělení a tudíž bychom chtěli vědět, zda náš kód dává očekávané výsledky při stejných znaménkách děleného a dělitele, při odlišných znaménkách a za situace, kdy je dělitel roven nule. Jeden ze scénářů nefunguje, tudíž upravíme funkci a scénář znovu vyzkoušíme. Nicméně je třeba vyzkoušet i scénáře ostatní - nemůžeme si být totiž jisti, zda jsme naší opravou nerozbili jinou funkčnost.  \n",
    "Nyní si představme, že máme v projektu mnoho funkcí a každá z nich se musí ověřit ve více než třech scénářích. Za takovýchto podmínek už na ruční testování není čas nemluvě o tom, že bychom mohli omylem na některý ze scénářů zapomenout. Proto by bylo vhodné, kdyby existovala předpřipravená množina scénářů, které bychom po každé změně mohli najednou pustit a ověřit, že se nám nic nepodařilo rozbít. Tímto opouštíme hájemství manuálního testování a dostáváme se k testování automatickému.  \n",
    "Testů existuje několikero druhů. My v kontextu tohoto povídání budeme řešit pouze testy jednotkové (unit testy). Jedná se o testy, které se omezují na krátký segment kódu - typicky funkce. V konstrastu k tomu lze postavit testy integrační, které testují celý běh programu/servisy.  \n",
    "Testy krom svého primárního smyslu přinášejí i pár dalších benefitů. Jejich psaní nutí programátora, aby testované funkce byly co možná nejmenší a nejjednodušší, ideálně bez vedlejších produktů. Díky tomu roste přehlednost kódu. Z pohledu na test člověk také pozná, jak vlastně vypadá vstup a výstup funkce, testy tak fungují jako doplněk dokumentace.  \n",
    "V Pythonu existuje několik frameworků na psaní testů. Ačkoli jeden z nich - unittest- je součástí základní instalace Pythonu, my si zde ukážeme práci s *pytestem*. Ten je totiž snazší na použití a hádám že i na porozumnění."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# První testy\n",
    "Mějme následující skript (k nalezení v adresáří first_script; podobně naleznete i ukázky pro další kapitoly v odpovídajících adresářích):\n",
    "```python\n",
    "def get_sum_of_two_numbers(first_number:int, second_number:int)->int:\n",
    "    return first_number + second_number\n",
    "\n",
    "def get_difference_of_two_numbers(first_number:int, second_number:int)->int:\n",
    "    return first_number - second_number\n",
    "```\n",
    "Jak bychom otestovali jeho první funkci *get_sum_of_two_numbers*? Nejprve si vytvoříme adresář, kde budeme mít testy umístěné. Na jméně z hlediska pytestu nezáleží, ale bývá zvykem, že nese jméno *tests* (v našem případě first_tests\\tests). V tomto adresáři vytvoříme soubor *test_get_sum_of_two_numbers.py*, v němž budou testy spojené s danou funkcí. Soubor by měl začínat s prefixem test\\_ (resp. se sufixem \\_test), aby byl pytestem při spouštění více testů najednou nalezen.  \n",
    "Soubor s jedním jednoduchým testem bude vypadat takto:\n",
    "```python\n",
    "import math_script\n",
    "\n",
    "def test_expected_sum_returned():\n",
    "    first_number = 10\n",
    "    second_number = 32\n",
    "    expected_result = 42\n",
    "    actual_result = math_script.get_sum_of_two_numbers(first_number, second_number)\n",
    "    assert expected_result == actual_result\n",
    "```\n",
    "Nejprve si importujeme testovaný skript. Následně se objeví testovací funkce mající v názvu prefix *test\\_* (bez toho by pytest nebral funkci jako testovací). V této funkci specifikujeme pro zadané vstupní parametry očekávaný výstup testované funkce (proměnná *expected_result*). Dále pak funkci s oněmi parametry provoláme a tak získáme výstup faktický (proměnná *actual_result*). Očekávaný a faktický výstup nakonec porovnáme v *assert* konstrukci. Tato konstrukce v případě, že je podmínka v ní platná, nedělá nic. Pokud je ale podmínka neplatná, bude vyvolána assert chyba (AssertionError). \n",
    "\n",
    "Test máme tedy napsaný a nyní bychom ho rádi pustili. Aby to fungovalo (aby byl test-skript test_get_sum_of_two_numbers.py schopen nalézt testovaný skript), musíme být v adresáři s testovaným skriptem - do něj se dostaneme s pomocí příkazu *cd*. Poté v konzoli spustíme\n",
    "```\n",
    "python -m pytest tests\\test_get_sum_of_two_numbers.py\n",
    "```\n",
    "Zde pythonu parametrem **-m** říkáme, že má balíček pytest pustit jako obyčejný skript. Poslední parametr pak představuje jméno testu.  \n",
    "Test by měl úspěšně proběhnout, tj. měli bychom vidět hlášku typu\n",
    "```\n",
    "================================================= test session starts =================================================\n",
    "platform win32 -- Python 3.7.4, pytest-7.1.2, pluggy-1.0.0\n",
    "rootdir: C:\\workshopy\\repozitar\\pytest\\first_tests\n",
    "plugins: mock-3.7.0\n",
    "collected 1 item\n",
    "\n",
    "tests\\test_get_sum_of_two_numbers.py .                                                                           [100%]\n",
    "\n",
    "================================================== 1 passed in 0.03s ==================================================\n",
    "```\n",
    "Obvykle ale chceme spustit více testů naráz. V konzolovém příkazu tak jako poslední parametr nezadáme jméno testu, ale jméno adresáře s testy:\n",
    "```\n",
    "python -m pytest tests\n",
    "```\n",
    "Po spuštění výše zmíněného uvidíme výstup, který už není tak optimistický jako výstup předešlý:\n",
    "```\n",
    "================================================= test session starts =================================================\n",
    "platform win32 -- Python 3.7.4, pytest-7.1.2, pluggy-1.0.0\n",
    "rootdir: C:\\workshopy\\repozitar\\pytest\\first_tests\n",
    "plugins: mock-3.7.0\n",
    "collected 3 items\n",
    "\n",
    "tests\\test_get_difference_of_two_numbers.py .F                                                                   [ 66%]\n",
    "tests\\test_get_sum_of_two_numbers.py .                                                                           [100%]\n",
    "\n",
    "====================================================== FAILURES =======================================================\n",
    "_________________________________________________ test_this_fill_fail _________________________________________________\n",
    "\n",
    "    def test_this_fill_fail():\n",
    ">       assert 1 == 2\n",
    "E       assert 1 == 2\n",
    "\n",
    "tests\\test_get_difference_of_two_numbers.py:11: AssertionError\n",
    "=============================================== short test summary info ===============================================\n",
    "FAILED tests/test_get_difference_of_two_numbers.py::test_this_fill_fail - assert 1 == 2\n",
    "============================================= 1 failed, 2 passed in 0.37s =============================================\n",
    "```\n",
    "To je dané tím, že v jednom z testů bylo natvrdo napsáno **assert 1 == 2**, což automaticky vedlo k jeho selhání. V první části reportu vidíme, u kterého souboru s testy nastaly problémy a v kolika testech, v druhé části jsou pak ony problémy explicitně ukázané.  \n",
    "Je možné, že při spuštění testů se něco nepovede. Následující výpis uvidíte v případě, že ve Vámi specifikovaném adresáři nejsou testy žádné, resp. zapomněli jste na prefix *test\\_* u testových souborů:\n",
    "```\n",
    "(environment) C:\\workshopy\\repozitar\\pytest\\first_tests>python -m pytest tests\n",
    "================================================= test session starts =================================================\n",
    "platform win32 -- Python 3.7.4, pytest-7.1.2, pluggy-1.0.0\n",
    "rootdir: C:\\vs\\programovani\\python\\workshopy\\repozitar\\pytest\\first_tests\n",
    "plugins: mock-3.7.0\n",
    "collected 0 items\n",
    "\n",
    "================================================ no tests ran in 0.00s ================================================\n",
    "```\n",
    "Za situace, kdy se budete snažit pouštět testy z adresáře, kde nejsou testované skripty, uvidíte hlášení\n",
    "```\n",
    "(environment) C:\\workshopy\\repozitar\\pytest>python -m pytest first_tests\\tests\\test_get_sum_of_two_numbers.py\n",
    "================================================= test session starts =================================================\n",
    "platform win32 -- Python 3.7.4, pytest-7.1.2, pluggy-1.0.0\n",
    "rootdir: C:\\workshopy\\repozitar\\pytest\n",
    "plugins: mock-3.7.0\n",
    "collected 0 items / 1 error\n",
    "\n",
    "======================================================= ERRORS ========================================================\n",
    "__________________________ ERROR collecting first_tests/tests/test_get_sum_of_two_numbers.py __________________________\n",
    "ImportError while importing test module 'C:\\workshopy\\repozitar\\pytest\\first_tests\\tests\\test_get_sum_of_two_numbers.py'.\n",
    "Hint: make sure your test modules/packages have valid Python names.\n",
    "Traceback:\n",
    "C:\\Python\\Python37\\lib\\importlib\\__init__.py:127: in import_module\n",
    "    return _bootstrap._gcd_import(name[level:], package, level)\n",
    "first_tests\\tests\\test_get_sum_of_two_numbers.py:1: in <module>\n",
    "    import math_script\n",
    "E   ModuleNotFoundError: No module named 'math_script'\n",
    "=============================================== short test summary info ===============================================\n",
    "ERROR first_tests/tests/test_get_sum_of_two_numbers.py\n",
    "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
    "================================================== 1 error in 0.08s ===================================================\n",
    "```\n",
    "Testování balíčků je principiálně stejné, jen si člověk musí dát pozor na to, aby se testy dostaly k testovaným funkcím. Ukázku nalezneme v adresáři *module_tests*.  \n",
    "Nakonec zmiňme tu největší zradu. Může se stát, že se nám při puštění testů spustí i část testovaného skriptu. K tomu dojde v tom případě, kdy v onom skriptu máme volně položený kód, tj. kód, který není ve funkci či třídě. Fakticky je toto chování dané importováním, které se musí v každém testovacím skriptu objevit. Řešení spočívá v umístění těchto \"volných\" příkazů do konstrukce *if \\_\\_name\\_\\_ == \"\\_\\_main\\_\\_\"*. V proměnné \\_\\_name\\_\\_ se totiž objeví jméno skriptu, pokud byl onen skript importován, anebo \"\\_\\_main\\_\\_\", pokud byl skript spuštěn napřímo."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mírně složitější testy\n",
    "Kontrola na rovnost celých čísel či textových řetězců patří k tomu nejjednoduššímu, co se dá testovat. V běžném provozu ale budeme muset být schopni ošetřit i komplikovanější situace. Těm se věnuje právě tato kapitola.  \n",
    "Typickým problémem je ověření toho, zda funkce vracející reálné číslo dělá to, co by dělat měla.\n",
    "```python\n",
    "def get_slightly_increased_number(orig_number):\n",
    "    increment = 0.001\n",
    "    return orig_number + increment\n",
    "```\n",
    "Pomiňme nyní skutečnost, že u této funkce je cifra vyčíslitelná a tak bychom si vystačili s rovností (ale ani toto neni samospásné - zkusme si spustit *assert 0.1 + 0.2 == 0.3*). Představme si namísto toho, že výsledek známe jen přibližně (jako by se dejme tomu jednalo o výpočet funkce sinus). Byli bychom spokojení, kdyby se výstup funkce rovnal očekávanému výstupu plus mínus nějaká malá hodnota. To bychom mohli v testu zakódovat jako dvojici assertů. Nicméně vhodnější (čti elegantnější) bude použít pytestovou funkci *approx*:\n",
    "```python\n",
    "import slightly_complicated\n",
    "import pytest\n",
    "\n",
    "def test_incrase_works():\n",
    "    orig_number = 5\n",
    "    expected_result = 5.001\n",
    "    actual_result = slightly_complicated.get_slightly_increased_number(orig_number)\n",
    "    \n",
    "    assert actual_result == pytest.approx(expected_result, abs=0.0005)\n",
    "\n",
    "def test_this_will_fail():\n",
    "    orig_number = 5\n",
    "    expected_result = 5.002\n",
    "    actual_result = slightly_complicated.get_slightly_increased_number(orig_number)\n",
    "    \n",
    "    assert actual_result == pytest.approx(expected_result, abs=0.0005)\n",
    "```\n",
    "Vidíme, že krom importu testovaného skriptu se musí importovat i samotný pytest. V testech samotných se obvyklým způsobem získá výstup testované funkce. V rámci assertu se ale tento výstup neporovnává s očekávaným výstupem, ale s výstupem funkce pytest.approx. Ta je krmena jednak očekávaným výstupem, jednak (volitelně) velikostí tolerance v parametru *abs*.  \n",
    "\n",
    "Někdy je vhodné netestovat hodnotu výstupu, ale jeho datový typ. To se provede pomocí funkce *isinstance*:\n",
    "```python\n",
    "import slightly_complicated\n",
    "\n",
    "def test_increased_number_is_float():\n",
    "    orig_number = 5\n",
    "    actual_result = slightly_complicated.get_slightly_increased_number(orig_number)\n",
    "    assert isinstance(actual_result, float)\n",
    "    \n",
    "def test_increased_number_is_integer():\n",
    "    orig_number = 5\n",
    "    actual_result = slightly_complicated.get_slightly_increased_number(orig_number)\n",
    "    assert isinstance(actual_result, int)\n",
    "```\n",
    "\n",
    "Když chceme otestovat, zda funkce vrací správný list či slovník, použijeme assertování a rovnosti, jako jsme aplikovali například u celých čísel. U dataframů to tak jednoduše nepůjde, tehdy pro test\n",
    "```python\n",
    "def test_naive_expected_frame():\n",
    "    expected_result = pd.DataFrame({\n",
    "        \"column_1\": [10, 20, 30],\n",
    "        \"column_2\": [100, 200, 300]\n",
    "    })\n",
    "    actual_result = slightly_complicated.get_some_frame()\n",
    "    assert expected_result == actual_result\n",
    "```\n",
    "obdržíme výstup\n",
    "```\n",
    "_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _\n",
    "\n",
    "self =    column_1  column_2\n",
    "0      True      True\n",
    "1      True      True\n",
    "2      True      True\n",
    "\n",
    "    def __nonzero__(self):\n",
    "        raise ValueError(\n",
    ">           f\"The truth value of a {type(self).__name__} is ambiguous. \"\n",
    "            \"Use a.empty, a.bool(), a.item(), a.any() or a.all().\"\n",
    "        )\n",
    "E       ValueError: The truth value of a DataFrame is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().\n",
    "\n",
    "..\\..\\..\\environment\\lib\\site-packages\\pandas\\core\\generic.py:1330: ValueError\n",
    "```\n",
    "V takovémto případě se musí použít funkce *assert_frame_equal* z *pandas.testing*. Konkrétně se výstup této funkce musí assertovat s předpokladem, že je roven None:\n",
    "```python\n",
    "import pandas as pd\n",
    "\n",
    "def test_expected_frame():\n",
    "    expected_result = pd.DataFrame({\n",
    "        \"column_1\": [10, 20, 30],\n",
    "        \"column_2\": [100, 200, 300]\n",
    "    })\n",
    "    actual_result = slightly_complicated.get_some_frame()\n",
    "    frames_differences = pd.testing.assert_frame_equal(expected_result, actual_result)\n",
    "    assert frames_differences is None\n",
    "```\n",
    "\n",
    "Za určitých okolností bývá užitečné ověřit, že jsou vytvořeny výjimky, když vytvořeny být mají. Tehdy musíme provolání testované funkce umístit do with konstrukce. V té bude namísto typického *open* funkce *pytest.raises*, která bude mít v argumentu výjimku, která by měla být vytvořena.\n",
    "```python\n",
    "import pytest\n",
    "\n",
    "def test_exception_raised():\n",
    "    first_number = 10\n",
    "    second_number = 0\n",
    "    with pytest.raises(slightly_complicated.UselessZeroDivisionException):\n",
    "        slightly_complicated.get_division_result(first_number, second_number)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pokročilé možnosti operativy s pytestem\n",
    "Při defaultním spouštění testů pomocí \n",
    "```\n",
    "python -m pytest tests\n",
    "```\n",
    "obdržíme v případě úspěšně proběhlých testů pouze informaci o tom, že z testového skriptu X úspěšně doběhlo Y testů\n",
    "```\n",
    "tests\\test_get_increment_by_five.py .......                                                                      [100%]\n",
    "```\n",
    "Níže budeme určité testy vyřazovat a tehdy by se hodilo vědět, které testy fakticky proběhly. To uvidíme jen tehdy, když budeme vyžadovat verbose výstup, což se zařídí přepínačem *-v*.\n",
    "```\n",
    "python -m pytest tests -v\n",
    "```\n",
    "Výstup pak bude mít podobu \n",
    "```\n",
    "tests/test_get_increment_by_five.py::test_simple_example_1 PASSED                                                [ 14%]\n",
    "tests/test_get_increment_by_five.py::test_simple_example_2 PASSED                                                [ 28%]\n",
    "tests/test_get_increment_by_five.py::test_simple_exercise PASSED                                                 [ 42%]\n",
    "tests/test_get_increment_by_five.py::test_trivial_exercise PASSED                                                [ 57%]\n",
    "tests/test_get_increment_by_five.py::test_marked_as_computation PASSED                                           [ 71%]\n",
    "tests/test_get_increment_by_five.py::test_marked_as_something PASSED                                             [ 85%]\n",
    "tests/test_get_increment_by_five.py::test_marked_as_computation_too PASSED                                       [100%]\n",
    "```\n",
    "Ještě důležitější může být verbosita u zfailovaných testů. Pro neverbózní provolání dostaneme například\n",
    "```\n",
    "_________________________________________________ test_this_will_fail _________________________________________________\n",
    "\n",
    "    def test_this_will_fail():\n",
    "        expected_result =  [\"three\"]\n",
    "        actual_result = some_tested_script.get_some_list()\n",
    "\n",
    ">       assert expected_result == actual_result\n",
    "E       AssertionError: assert ['three'] == ['one', 'two', 'three']\n",
    "E         At index 0 diff: 'three' != 'one'\n",
    "E         Right contains 2 more items, first extra item: 'two'\n",
    "E         Use -v to get more diff\n",
    "\n",
    "tests\\test_get_some_list.py:7: AssertionError\n",
    "```\n",
    "Při verbózním provoláním už ale vidíme\n",
    "```\n",
    "_________________________________________________ test_this_will_fail _________________________________________________\n",
    "\n",
    "    def test_this_will_fail():\n",
    "        expected_result =  [\"three\"]\n",
    "        actual_result = some_tested_script.get_some_list()\n",
    "\n",
    ">       assert expected_result == actual_result\n",
    "E       AssertionError: assert ['three'] == ['one', 'two', 'three']\n",
    "E         At index 0 diff: 'three' != 'one'\n",
    "E         Right contains 2 more items, first extra item: 'two'\n",
    "E         Full diff:\n",
    "E         - ['one', 'two', 'three']\n",
    "E         + ['three']\n",
    "\n",
    "tests\\test_get_some_list.py:7: AssertionError\n",
    "```\n",
    "Pro komplikovanější případy lze použít i very verbose režím, kdy se použije přepínač *-vv*. Tehdy už není schována opravdu žádná část výstupu.\n",
    "\n",
    "\n",
    "Výše jsme viděli, jak spouštět testy buď explicitně po jednom, anebo všechny najednou. Občas se ale hodí mít něco mezi zmíněnými extrémy. Dejme tomu, že při větší změně programu rozbijeme několik funkcionalit. V rámci testů ale chceme napřed ověřit, že funguje featura A, přičemž nechceme ztrácet čas testováním věcí, o kterých víme, že automaticky zfailují.  \n",
    "Pakliže chceme spustit pouze testy obsahující v názvu nějaké slovo, předáme toto slovo jako doplněk k parametru *-k*. Tj. například pokud chceme spustit testy obsahující \"example\", spustíme v konzoli\n",
    "```\n",
    "python -m pytest -v -k example tests\n",
    "```\n",
    "Všimněte si, že na konci konzolového výstupu bude napsáno\n",
    "```\n",
    "================== X passed, Y deselected in 0.19s ==========\n",
    "```\n",
    "Tj. explicitně vidíme, kolik testů bylo přeskočeno.  \n",
    "S podmínkami na (ne)přítomnost slov si můžeme pohrát i více. Můžeme například chtít pustit všechny testy, které zvolené slovo v názvu *neobsahují*. Pro to musíme vložit do příkazu před ono slovo *not*, avšak současně musíme celou konstrukci umístit do uvozovek:\n",
    "```\n",
    "python -m pytest -v -k \"not example\" tests\n",
    "```\n",
    "Dále můžeme požadovat přítomnost více slov, kteréžto v podmínce spojíme pomocí *and*:\n",
    "```\n",
    "python -m pytest -v -k \"simple and exercise\" tests\n",
    "```\n",
    "Případně můžeme použít i logiku typu *or*:\n",
    "```\n",
    "python -m pytest -v -k \"simple or exercise\" tests\n",
    "```\n",
    "Testy můžeme už v testovacím skriptu sdružovat do kategorií a to prostřednictvím custom markerů (fakticky se jedná o dekorátory). Zdůrazněme, že v takomém případě je opět nutné importovat v testovacím skriptu pytest:\n",
    "```\n",
    "import pytest\n",
    "\n",
    "@pytest.mark.computation\n",
    "def test_marked_as_computation():\n",
    "    orig_number = 1\n",
    "    expected_result = 6\n",
    "    actual_result = some_tested_script.get_increment_by_five(orig_number)\n",
    "    assert expected_result == actual_result\n",
    "```\n",
    "Spuštění testů s daným markerem se poté zařídí parametrem *-m* následovaným názvem markeru\n",
    "```\n",
    "python -m pytest -v -m computation\n",
    "```\n",
    "Opět lze aplikovat rudimentální and/or/not logiku:\n",
    "```\n",
    "python -m pytest -v -m \"not computation\"\n",
    "```\n",
    "Pokud jsme kódy nestáhli z repozitáře, můžeme si všimnout, že se nám objevují warningy spojené s tím, že custom markery nejsou registrované:\n",
    "```\n",
    "PytestUnknownMarkWarning: Unknown pytest.mark.computation - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html\n",
    "```\n",
    "Pokud bychom chtěli warningy potlačit, použijeme parametr *-disable-warnings*\n",
    "```\n",
    "python -m pytest -v -m computation --disable-warnings\n",
    "```\n",
    "Nicméně vhodnější bude se s warningy vypořádat. V tomto kontextu to znamená custom markery zaregistrovat. To v praxi znamená, že v adresáři, ze kterého testy spouštíme, vyrobíme soubor *pytest.ini* o následujícím obsahu:\n",
    "```\n",
    "[pytest]\n",
    "markers =\n",
    "    computation: some label\n",
    "    something\n",
    "```\n",
    "Sekce za dvojtečkou je nepovinný popisek daného markeru.  \n",
    "\n",
    "Doposud jsme mluvili o custom markerech, existují ale i \"normální\" markery. Například *skip* způsobí vynechání testu.  Marker \"xfail\" způsobí, že test sice proběhne, ale i pokud selže, nedostane se do výsledných statistik jako \"nomrálně\" zfailnuté testy.\n",
    "```\n",
    "@pytest.mark.skip    \n",
    "def test_to_skip():\n",
    "    orig_number = 1\n",
    "    expected_result = 6\n",
    "    actual_result = some_tested_script.get_increment_by_five(orig_number)\n",
    "    assert expected_result == actual_result\n",
    "\n",
    "@pytest.mark.xfail\n",
    "def test_to_maybe_fail():\n",
    "    orig_number = 1\n",
    "    expected_result = 6\n",
    "    actual_result = some_tested_script.get_increment_by_five(orig_number)\n",
    "    assert expected_result == actual_result    \n",
    "    \n",
    "@pytest.mark.xfail\n",
    "def test_to_fail():\n",
    "    orig_number = 1\n",
    "    expected_result = 7\n",
    "    actual_result = some_tested_script.get_increment_by_five(orig_number)\n",
    "    assert expected_result == actual_result\n",
    "```\n",
    "Na výstupu totiž uvidíme\n",
    "```\n",
    "tests/test_get_increment_by_five.py::test_to_skip SKIPPED (unconditional skip)                                   [ 72%]\n",
    "tests/test_get_increment_by_five.py::test_to_maybe_fail XPASS                                                    [ 81%]\n",
    "tests/test_get_increment_by_five.py::test_to_fail XFAIL                                                          [ 90%]\n",
    "```\n",
    "a\n",
    "```\n",
    "=========== 2 failed, 6 passed, 1 skipped, 1 xfailed, 1 xpassed in 0.92s ===========\n",
    "```\n",
    "Asi nejzajímavější z out-of-the-box markerů je ale *parametrize*. Ten dovoluje efektivně v rámci jedné definice testu vytvořit několik testů naráz, kteréžto testy se ale v rámci vyhodnocení berou jako samostatné jednotky. Jako v předchozích případech se před testovací funkci napíše dekorátor markeru. Ten bude ale tentokrát následován kulatými závorkami s parametry. Prvním parametrem bude textový řetězec, ve kterém čárky oddělují názvy proměnných, kterými bude testovací funkce krmena. Druhý parametr tvoří list tuplů. Ty budou odpovídat názvům proměnným z parametru prvního. Samotná testovací funkce pak musí parametry přebírat a ve svých vnitřnostech je nějak zpracovávat. Výsledný produkt tedy vypadá takto:\n",
    "```\n",
    "@pytest.mark.parametrize(\"some_input, expected_result\", [(11, 16), (20, 25), (20, 26)])   \n",
    "def test_parametrized(some_input, expected_result):\n",
    "    actual_result = some_tested_script.get_increment_by_five(some_input)\n",
    "    assert expected_result == actual_result\n",
    "```\n",
    "Ve výstupu pytestu je opravdu tato jedna testovací funkce reprezentována, jako by šlo o tři nezávislé testy:\n",
    "```\n",
    "tests/test_get_increment_by_five.py::test_parametrized[11-16] PASSED                                             [ 78%]\n",
    "tests/test_get_increment_by_five.py::test_parametrized[20-25] PASSED                                             [ 85%]\n",
    "tests/test_get_increment_by_five.py::test_parametrized[20-26] FAILED                                             [ 92%]\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mocky\n",
    "Občas se do testované funkce předává nějaký objekt. Na něm by se ve funkci měly provést určité operace. No a my bychom chtěli otestovat, že k nim opravdu dojde. Jenže co když není reálné, abychom v rámci testu reálný objekt vytvořili, například poněvadž se jedná o connectionu do produkční databáze? Tehdy se hodí použít mocky, tj. fakeové objekty, které namísto původní funkčnosti dokážou sledovat, jak často a s jakými parametry na nich byla ta která metoda provolána.  \n",
    "Nechť testovaná funkce přebírá pandí dataframe a dvakrát do něj přidá sloupec pomocí metody insert\n",
    "```python\n",
    "import pandas as pd\n",
    "\n",
    "def do_something_with_frame(orig_frame:pd.DataFrame)->None:\n",
    "    orig_frame.insert(1, \"some_column\", 42)\n",
    "    orig_frame.insert(1, \"another_column\", 142)\n",
    "```\n",
    "Nyní v testu nebudeme do testované funkce vkládat reálný dataframe, nýbrž mock. Ten získáme z \"původního\" pythoního testovacího balíčku *unittest*, nikoli z pytestu. Vytvoříme ho provolání třídy *Mock* a následně zavoláme testovanou funkci. Ta nám sice dataframe/mock nevrací zpátky, ale to nevadí. Změny objektu vlivem provolání inserty se totiž díky předávání reference a nikoli dat dostanou ven z funkce.  \n",
    "Chtějme nejprve otestovat, že metoda insert byla zavolána dvakrát. K tomu se dostaneme tak, že se na mockovaném objektu skrz tečkovou notaci zeptáme na *insert* (vytvořené v okamžik provolání stejnojmenou metodou) a u něj na parametr *call_count*. Následuje už obyčejné assertění\n",
    "```python\n",
    "def test_insert_called_two_times():\n",
    "    mocked_object = Mock()\n",
    "    script_for_mocks.do_something_with_frame(mocked_object)\n",
    "    expected_insert_calls = 2\n",
    "    actual_insert_calls = mocked_object.insert.call_count\n",
    "    \n",
    "    assert expected_insert_calls == actual_insert_calls\n",
    "```\n",
    "Častokrát se určitá metoda provolá jen jednou. Proto byla na mockovacím objektu vytvořena metoda *assert_called_one*. V našem připadě jsme metodu provolali dvakrát, tudíž následující test failne.\n",
    "```python\n",
    "def test_insert_called_once():\n",
    "    mocked_object = Mock()\n",
    "    script_for_mocks.do_something_with_frame(mocked_object)\n",
    "    \n",
    "    mocked_object.insert.assert_called_once()\n",
    "```\n",
    "Pokud potřebujeme pouze zkontrolovat, zda metoda provolaná byla (a je nám jedno kolikrát), použijeme *assert_called*.\n",
    "```python\n",
    "def test_insert_called():\n",
    "    mocked_object = Mock()\n",
    "    script_for_mocks.do_something_with_frame(mocked_object)\n",
    "    \n",
    "    mocked_object.insert.assert_called()\n",
    "    \n",
    "def test_drop_called():\n",
    "    mocked_object = Mock()\n",
    "    script_for_mocks.do_something_with_frame(mocked_object)\n",
    "    \n",
    "    mocked_object.drop.assert_called()\n",
    "```\n",
    "Občas bývá potřeba ověřit, že metoda byla provolaná s očekávanými parametry. Pokud nám jde jen o parametry posledního provolání (resp. pokud si jsme jistí, že k více než jednomu provolání nedojde), použijeme *call_args*. Jedná se o call objekt (pro naše obvyklé účely efektivně tuple), který má dvě složky. V první se ukládají args, v druhé kwargs.\n",
    "```python\n",
    "def test_insert_called_last_time_with_expected_params():\n",
    "    mocked_object = Mock()\n",
    "    script_for_mocks.do_something_with_frame(mocked_object)\n",
    "    expected_params = (1, \"another_column\", 142)\n",
    "    #because call_args is a call object which have args tuple at index 0\n",
    "    actual_params = mocked_object.insert.call_args[0]\n",
    "    assert expected_params == actual_params\n",
    "```\n",
    "Pokud je nutné prověřit argumenty všech volání, musí se sáhnout po *call_args_list*, což je list call objektů.\n",
    "```python\n",
    "def test_insert_called_both_times_with_expected_params():\n",
    "    mocked_object = Mock()\n",
    "    script_for_mocks.do_something_with_frame(mocked_object)\n",
    "    \n",
    "    first_expected_params = (1, \"some_column\", 42)\n",
    "    second_expected_params = (1, \"another_column\", 142)\n",
    "\n",
    "    first_call_params, second_call_params = mocked_object.insert.call_args_list\n",
    "    first_actual_params = first_call_params[0]\n",
    "    second_actual_params = second_call_params[0]\n",
    "    assert first_expected_params == first_actual_params\n",
    "    assert second_expected_params == second_actual_params\n",
    "```\n",
    "V případě, že chceme ověřit provolání args i kwargs, tj. namísto původní testované funkce máme třeba\n",
    "```python\n",
    "def do_something_with_frame_args_kwargs(orig_frame:pd.DataFrame)->None:\n",
    "    orig_frame.insert(1, \"some_column\", value=42)\n",
    "    orig_frame.insert(1, \"another_column\", value=142)\n",
    "```\n",
    "budeme využívat i druhou složku call_args, tj. test se bude nést v následujícím duchu:\n",
    "```python\n",
    "import script_for_mocks\n",
    "from unittest.mock import Mock\n",
    "\n",
    "def test_insert_called_last_time_with_expected_params():\n",
    "    mocked_object = Mock()\n",
    "    script_for_mocks.do_something_with_frame_args_kwargs(mocked_object)\n",
    "    expected_args = (1, \"another_column\")\n",
    "    expected_kwargs = {\"value\":142}\n",
    "    \n",
    "    actual_args = mocked_object.insert.call_args[0]\n",
    "    actual_kwargs = mocked_object.insert.call_args[1]\n",
    "    \n",
    "    assert expected_args == actual_args\n",
    "    assert expected_kwargs == actual_kwargs\n",
    "```\n",
    "Výše jsme mocky použily jako svého druhu sondu do funkce. Nicméně mockům lze i říci, že se mají nějak chovat. Například lze zařídit, aby jejich metody vracely věci, které chceme. Mějme testovanou funkci \n",
    "```python\n",
    "def get_frame_without_col(orig_frame:pd.DataFrame, col_name:str)->pd.DataFrame:\n",
    "    something = orig_frame.drop(column=col_name)\n",
    "    return something\n",
    "```\n",
    "Normálně by *drop* v této funkci vracel dataframe. Můžeme ale zařídit, aby v rámci testu vracel string a to sice pomocí *mocked_object.drop.return_value*:\n",
    "```python\n",
    "def test_insert_in_mock_returns_something():\n",
    "    mocked_object = Mock()\n",
    "    mocked_object.drop.return_value = \"there should be a frame\"\n",
    "    returned_value = script_for_mocks.get_frame_without_col(mocked_object, \"some_col\")\n",
    "    \n",
    "    assert returned_value == \"there should be a frame\"\n",
    "```\n",
    "\n",
    "Pro úplnost dodejme, že krom třídy Mock existuje i třída MagicMock. Ta se od svého předchodce liší tím, že má zadefinované základní magické (dunder; podtržítkové) metody. Ty se u Mocku narozdíl od metod normálích při provolání nevytváří, namísto toho program zahlásí error."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "<MagicMock name='mock.__getitem__()' id='2265866571144'>\n",
      "[]\n",
      "True\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "object of type 'Mock' has no len()",
     "output_type": "error",
     "traceback": [
      "\u001B[1;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[1;31mTypeError\u001B[0m                                 Traceback (most recent call last)",
      "\u001B[1;32m<ipython-input-12-d9fef771ea94>\u001B[0m in \u001B[0;36m<module>\u001B[1;34m\u001B[0m\n\u001B[0;32m     11\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m     12\u001B[0m \u001B[1;31m#operations using dunder methods lead to errors on simple Mock objects\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m---> 13\u001B[1;33m \u001B[0mlen\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0msimple_mock\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m",
      "\u001B[1;31mTypeError\u001B[0m: object of type 'Mock' has no len()"
     ]
    }
   ],
   "source": [
    "from unittest.mock import Mock, MagicMock\n",
    "\n",
    "simple_mock = Mock()\n",
    "magical_mock = MagicMock()\n",
    "\n",
    "#operations using dunder methods work on MagicMock objects\n",
    "print(len(magical_mock))\n",
    "print(magical_mock[0])\n",
    "print(list(magical_mock))\n",
    "print(magical_mock != 2)\n",
    "\n",
    "#operations using dunder methods lead to errors on simple Mock objects\n",
    "len(simple_mock)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Freezegun\n",
    "Práce s datumem a časem bývá obvykle problematická a testování není výjimkou. Namockování datumových objektů, zejména používá-li se timedelta, není úplně triviální. Proto se hodí, že byl pro tyto účely vyvinut speciální balíček - freezegun.\n",
    "Mějme testovanou funkci, která vrací datum sedm dní před svým spuštěním. Zdůrazněme, že aktuální datum do funkce jako parametr nevstupuje, nýbrž se určuje v jejím vnitřku.\n",
    "```python\n",
    "import datetime\n",
    "\n",
    "def get_date_week_ago():\n",
    "    today = datetime.date.today()\n",
    "    week_ago = today - datetime.timedelta(days=7)\n",
    "    return week_ago\n",
    "```\n",
    "Při testování by byl nyní problém, jak zafixovat funkci *today* na jeden konkrétní den. No a právě to nám zajistí freezegun. Co se samotné realizace v testu, existuje několik co do výsledku ekvivaletních přístupů. V prvním z nich nastavíme aktuální datum a čas pomocí *freezegun.freeze_time*. Nicméně čas bude za/nastaven až když se zavolá na odpovídajícím objektu metoda *start*. Reálný čas se vrátí po následném provolání metody *stop*.\n",
    "```python\n",
    "import freezegun\n",
    "import datetime\n",
    "import script_for_freezegun\n",
    "\n",
    "def test_raw_use_freezer():\n",
    "    expected_result = datetime.date(2022, 1, 21)\n",
    "    \n",
    "    freezer = freezegun.freeze_time(\"2022-01-28 12:00:01\")\n",
    "    freezer.start()\n",
    "    actual_result = script_for_freezegun.get_date_week_ago()\n",
    "    freezer.stop()\n",
    "    \n",
    "    assert expected_result == actual_result\n",
    "```\n",
    "Druhý přístup využívá kontextového manažeru (with konstrukce) - čas je zamrzlý jen v jeho rámci:\n",
    "```python\n",
    "def test_freezer_context_manager():\n",
    "    expected_result = datetime.date(2022, 1, 21)\n",
    "    \n",
    "    with freezegun.freeze_time(\"2022-01-28 12:00:01\"):\n",
    "        actual_result = script_for_freezegun.get_date_week_ago()\n",
    "    \n",
    "    assert expected_result == actual_result\n",
    "```\n",
    "Nakonec přístup třetí používá dekorátor, přičemž v čase je zamrzlá celá testovací funkce:\n",
    "```python\n",
    "@freezegun.freeze_time(\"2022-01-28 12:00:01\")\n",
    "def test_freezer_decorator():\n",
    "    expected_result = datetime.date(2022, 1, 21)\n",
    "    \n",
    "    actual_result = script_for_freezegun.get_date_week_ago()\n",
    "    \n",
    "    assert expected_result == actual_result\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Monkeypatching\n",
    "Při mockování si vyrábíme falešný objekt, který pak vložíme do testované funkce jako jeden z jejích paremtrů. Co když ale musíme zabránit volání metody na objektu, který vzniká *až* uvnitř funkce? Případně co když chceme nahradit funkci, která je volaná uvnitř testované funkce?  \n",
    "Jako příklad testované funkce si vezměme následující konstrukci:\n",
    "```python\n",
    "import pandas as pd\n",
    "\n",
    "def get_some_frame():\n",
    "    some_frame = pd.DataFrame({\n",
    "        \"first_col\": [1, 2, 3],\n",
    "        \"second_col\": [10, 20, 30]\n",
    "    })\n",
    "    \n",
    "    frame_from_method = some_frame.copy()\n",
    "\n",
    "    return frame_from_method \n",
    "```\n",
    "Nyní si představme, že chceme zabránit tomu, aby se při testování reálně metoda *copy* na dataframu zavolala. Řešením je použít monkeypatchování. Při něm si vytvoříme falešnou funkci, která je oproti originálu neškodná (nesnaží se nic provolávat, nesnaží se zapisovat do databáze apod.) a následně skutečnou funkci nahradíme touto falešnou. Jak tyto operace v reálném provedení vypadají? Nejprve musíme do parametrů testovací funkce napsat \"*monkeypatch*\" (fakticky se jedná o tzv. fixture, ale o tom až o pár kapitol dále). Následně si (ve scopu testovací funkce) vytvoříme falešnou copy funkci. Ta jako parametry přebírá \\*args a \\*\\*kwargs, abychom nemuseli řešit, co za parametry se fakticky pouští do reálné funkce v testované konstrukci. Zde vrací funkce *fake_copy* dataframe, často ale falešné funkce nevracejí nic a obsahují jediný příkaz - *pass*. Nahrazení pravé funkce za falešnou se provádí s pomoci *monkeypatch.setattr*. U něj je první parametr místo, ze kterého funkce pochází. Jelikož v příkladu řešíme metodu pandího dataframu, musí zde být *pd.DataFrame*. Zde se ono *pd* vztahuje k aliasu použitém v testovaném skriptu. Pokud by v něm namísto *import pandas as pd* bylo *import pandas*, museli bychom mít jako první parametr *setattr* *pandas.DataFrame*. Druhým parametrem *setattr* je název nahrazované funkce **ve stringové podobě**, tj. v uvozovkách. Nakonec třetím parametrem je nahrazující funkce (jako funkce, tj. bez uvozovek). Dále se v testu postupuje jako obvykle, tj. provoláme testovanou funkci a provedeme assert s použitím vráceného objektu.\n",
    "```python\n",
    "import script_for_monkeypatching\n",
    "import pandas as pd\n",
    "\n",
    "def test_something(monkeypatch):\n",
    "    def fake_copy(*args, **kwargs):\n",
    "        return pd.DataFrame({\"fake_column\": [42, 142]})\n",
    "        \n",
    "    monkeypatch.setattr(pd.DataFrame, \"copy\", fake_copy)\n",
    "    \n",
    "    result_frame = script_for_monkeypatching.get_some_frame()\n",
    "    \n",
    "    assert result_frame.columns == [\"fake_column\"]\n",
    "```\n",
    "Situace je lehce odlišná v přápadě, kdy v testovaném skriptu importujeme nahrazovanou funkci pomocí fromu:\n",
    "```python\n",
    "from numpy import sqrt\n",
    "\n",
    "def get_list_sqrt(input_list):\n",
    "    return sqrt(input_list) \n",
    "```\n",
    "Poté totiž z hlediska testování ona funkce nebydlí v nějakém balíčku, ale právě v testovaném skriptu. Prvním parametrem *setattr* tak musí být právě testovaný skript:\n",
    "```python\n",
    "import script_for_monkeypatching\n",
    "\n",
    "def test_fake_sqrt(monkeypatch):\n",
    "    def fake_sqrt(*args, **kwargs):\n",
    "        return [1, 1, 1, 1]\n",
    "        \n",
    "    monkeypatch.setattr(script_for_monkeypatching, \"sqrt\", fake_sqrt)\n",
    "    \n",
    "    result_list = script_for_monkeypatching.get_list_sqrt([2, 4, 9])\n",
    "    \n",
    "    assert [1, 1, 1, 1] == result_list\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Spy\n",
    "U mocků jsme ony falešné objekty nepoužívali pouze jako náhradu za objekty skutečné, ale též jako počítadlo, kolikrát na nich byla ta které metoda zavolaná. Stejná potřeba může vyvstat u metod používaných v testované funkci. Představme si například, že bychom u níže uvedené metody chtěli otestovat, zda byla dataframová metoda *insert* provolána v očekávaném počtu.\n",
    "```python\n",
    "def get_another_frame():\n",
    "    some_frame = pd.DataFrame({\n",
    "        \"first_col\": [1, 2, 3],\n",
    "        \"second_col\": [10, 20, 30]\n",
    "    })\n",
    "    \n",
    "    some_frame.insert(1, \"third_col\", [40, 50, 60])\n",
    "    some_frame.insert(1, \"fourth_col\", [400, 500, 600])\n",
    "    frame_from_method = some_frame.copy()\n",
    "\n",
    "    return frame_from_method \n",
    "```\n",
    "Abychom mohly počty provolání funkcí sledovat, musíme nainstalovat balíček pytest-mock. Poté do volání testovací funkce přidáme parametr *mocker*. Z něho si v těle testu vezmeme metodu *spy*, která přebírá jako své dva parametry to samé, co sypeme do prvních dvou parametrů *monkeypatch.setattr* - domov sledované funkce a její jméno coby string. Po provolání testované metody pak počet provolání dostaneme ze spy objektu, konkrétně z jeho atributu *call_count*.\n",
    "```python\n",
    "import script_for_spying\n",
    "import pandas as pd\n",
    "\n",
    "def test_insert_called_twice_at_real(mocker):\n",
    "\n",
    "    spy_insert = mocker.spy(pd.DataFrame, \"insert\")\n",
    "    \n",
    "    script_for_spying.get_another_frame()\n",
    "    \n",
    "    expected_call_count = 2\n",
    "    actual_call_count = spy_insert.call_count\n",
    "    \n",
    "    assert expected_call_count == actual_call_count\n",
    "```\n",
    "Výše uvedený příklad počítal s tím, že metoda insert nebude nahrazována. I kdyby ale s pomocí monkeypatche nahrazena byla, pro *spy* se nic nemění - i tehdy bude v jeho parametrech jméno původní funkce:\n",
    "```python\n",
    "import script_for_spying\n",
    "import pandas as pd\n",
    "\n",
    "def test_insert_called_twice_at_fake(monkeypatch, mocker):\n",
    "    def fake_insert(*args, **kwargs):\n",
    "        pass\n",
    "    \n",
    "    monkeypatch.setattr(pd.DataFrame, \"insert\", fake_insert)\n",
    "    spy_insert = mocker.spy(pd.DataFrame, \"insert\")\n",
    "    \n",
    "    script_for_spying.get_another_frame()\n",
    "    \n",
    "    expected_call_count = 2\n",
    "    actual_call_count = spy_insert.call_count\n",
    "    \n",
    "    assert expected_call_count == actual_call_count\n",
    "```\n",
    "Krom počtu provolání funkce můžeme i sledovat, s jakými parametry byla volaná, a to díky *assert_called_with*. \n",
    "```python\n",
    "import script_for_spying\n",
    "\n",
    "def test_inner_function_called_with_expected_params(mocker):\n",
    "    spy_get_nothing = mocker.spy(script_for_spying, \"get_nothing\")\n",
    "    \n",
    "    script_for_spying.wrapper_function()\n",
    "    \n",
    "    spy_get_nothing.assert_called_with(42, \"hello\")\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testování kontextových managerů\n",
    "Kontextové managery (typicky with open konstrukce) stojí na \\_\\_enter\\_\\_ a \\_\\_exit\\_\\_ metodách. Když se použije *with*, je zavolán *\\_\\_enter\\_\\_*, který vrací objekt použitý v kontextu. Pak se spustí vnitřní (odsazený) blok kódu. Metoda *\\_\\_exit\\_\\_* se zavolá na konec a to nezávisle na tom, jaký je výstup výše zmíněného bloku kódu. Tj. častokrát používaná jednoduchá konstrukce\n",
    "```python\n",
    "with some_context_function(something) as anything:\n",
    "    do_some_magic(anything)\n",
    "```\n",
    "je ekvivalentní této divočině:\n",
    "```python\n",
    "cont = some_context_function(something)\n",
    "anything = cont.__enter__()\n",
    "try:\n",
    "    do_some_magic(anything)\n",
    "finally:\n",
    "    cont.__exit__(None, None, None)\n",
    "```\n",
    "\n",
    "Z hlediska psaní testů pro open funkci je třeba vědět, že její definice vypadá následovně:\n",
    "```python\n",
    "def open(path):\n",
    "    f = builtins.open(path, \"r\")\n",
    "    return f\n",
    "```\n",
    "Tj. jedná se o funkci vracející objekt.\n",
    "\n",
    "Nnyí se podívejme na reálný příklad. Chtějme v testu ověřit, že v následující funkci dochází k provolání *write* na objektu *file*.\n",
    "```python\n",
    "def create_html_page():\n",
    "    page_text = \"<b>Hello world</b>\"\n",
    "    \n",
    "    with open(\"index.html\", \"w+\", encoding=\"utf-8\") as file:\n",
    "        file.write(page_text)\n",
    "```\n",
    "V testu musíme vytvořit fakevou funkci open. Výše vidíme, že tato funkce musí vracet objekt disponující metodami *_\\_enter\\_\\_* a *\\_\\_exit\\_\\_*. To  znamená, že *fake_open* musí vracet nikoli *Mock*, ale *MagicMock*. Nicméně ta věc, na které se bude provolávat *write* (a kterou musí vracet *\\_\\_enter\\_\\_* metoda MagicMocku), už výše uvedené speciální funkce nepotřebuje, může jí tudíž být normální Mock. Jelikož ale tento Mock potřebujeme v rámci testu, abychom z něj dostali parametr *call_count*, je třeba ho definovat nikoli ve scopu vnořené funkce *fake_open*, ale ve vnějším scopu testovací funkce.\n",
    "\n",
    "```python\n",
    "from unittest.mock import Mock, MagicMock\n",
    "import script_for_context\n",
    "\n",
    "def test_open_called_once(monkeypatch):\n",
    "    fake_file_object = Mock()\n",
    "    \n",
    "    def fake_open(*args, **kwargs):\n",
    "        context_object = MagicMock()\n",
    "        context_object.__enter__.return_value = fake_file_object\n",
    "        return context_object\n",
    "    \n",
    "    monkeypatch.setattr(\"builtins.open\", fake_open)\n",
    "    \n",
    "    script_for_context.create_html_page()\n",
    "\n",
    "    assert 1 == fake_file_object.write.call_count\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fixtures\n",
    "Mějme funkci\n",
    "```python\n",
    "import pandas as pd\n",
    "\n",
    "def get_frame_column_count_plus_one(input_frame:pd.DataFrame)->int:\n",
    "    input_frame.insert(1, \"some_col_name\", [10, 20, 30])\n",
    "    return input_frame.shape[1]\n",
    "```\n",
    "kterou testujeme pomocí\n",
    "```python\n",
    "import script_for_fixtures\n",
    "import pandas as pd\n",
    "\n",
    "def test_returned_expected_column_count_no_fixture():\n",
    "    dummy_frame = pd.DataFrame({\"some_col\":[1,2,3]})\n",
    "    \n",
    "    expected_result = 2\n",
    "    actual_result = script_for_fixtures.get_frame_column_count_plus_one(dummy_frame)\n",
    "    \n",
    "    assert expected_result == actual_result\n",
    "    \n",
    "def test_insert_called_once_no_fixture(mocker):\n",
    "    dummy_frame = pd.DataFrame({\"some_col\":[1,2,3]})\n",
    "    spy_insert = mocker.spy(pd.DataFrame, \"insert\")\n",
    "    \n",
    "    script_for_fixtures.get_frame_column_count_plus_one(dummy_frame)\n",
    "    \n",
    "    spy_insert.assert_called_once()\n",
    "```\n",
    "Což o to, tyto testy fungují. Problém ale spočívá ve skutečnosti, že si obě testovací funkce vytváří svůj vlastní dummy_frame. V tomto konkrétním případě je to jedno, avšak mohly by nastat situace, kdy by vytváření parametru testované funkce zabíralo celou obrazovku. Pak by se snadno mohlo stát, že vytváření parametru bude v testovacích funkcích nedopatřením odlišné a tak se testovací scénaře začnou chovat nekonzistentně. Proto by bylo užitečné mít možnost kód společný pro více testovacích funkcí vložit do funkce jiné - do tzv. fixture funkce.  \n",
    "Fixture funkce vypadá jako každá jiná, jen se nad jejím jménem nalézá dekorátor *pytest.fixture*. Do parametrů testovacích funkcí následně vkládáme její název. Dojde k tomu, že v těle testované funkce vznikne proměnná stejného jména, jaké nese fixture funkce. Přitom tato proměnná obsahuje věci vracené fixture funkcí. Tj. v našem konkrétním případě se v testovacích funkcích pod frame_for_test skrývá dataframe vrácený stejnojmennou funkcí. \n",
    "```python\n",
    "import script_for_fixtures\n",
    "import pandas as pd\n",
    "import pytest\n",
    "\n",
    "@pytest.fixture\n",
    "def frame_for_test()->pd.DataFrame:\n",
    "    return pd.DataFrame({\"some_col\":[1,2,3]})\n",
    "\n",
    "def test_returned_expected_column_count_used_fixture(frame_for_test):\n",
    "    expected_result = 2\n",
    "    actual_result = script_for_fixtures.get_frame_column_count_plus_one(frame_for_test)\n",
    "    \n",
    "    assert expected_result == actual_result\n",
    "    \n",
    "def test_insert_called_once_used_fixture(frame_for_test, mocker):\n",
    "\n",
    "    spy_insert = mocker.spy(pd.DataFrame, \"insert\")\n",
    "    \n",
    "    script_for_fixtures.get_frame_column_count_plus_one(frame_for_test)\n",
    "    \n",
    "    spy_insert.assert_called_once()\n",
    "```\n",
    "Takto jsme vyřešili problém zbytečného opakování stejného kódu. Co kdyby ale vytváření objektu ve fixture funkci bylo z hlediska hardwaru hodně náročné a současně se tento objekt v rámci testovacích funkcí neměnil? Pak by se nám defaultní chování fixture funkcí - separátní provolání pro každou testovacích funkcí - zrovna nehodilo. To lze ale napravit tím, že k dekorátoru samotnému přidáme parametr *scope=\"module\"*. Pak se totiž fixture zavolá jen jednou za celý testovací skript a objekt, který je takto vytvořen, bude použit všemi testovacími funkcemi v testovacím skriptu.\n",
    "```python\n",
    "@pytest.fixture(scope=\"module\")\n",
    "def frame_for_test()->pd.DataFrame:\n",
    "    return pd.DataFrame({\"some_col\":[1,2,3]})\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
